aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt

```diff
@@ -645,7 +646,72 @@ class AWSCognitoAuthPlugin : AuthPlugin<AWSCognitoAuthServiceBehavior>() {
         onSuccess: Consumer<MutableList<AuthDevice>>,
         onError: Consumer<AuthException>
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                is AuthenticationState.SignedIn -> {
+                    var token: StateChangeListenerToken? = null
+                    token = credentialStoreStateMachine.listen(
+                        {
+                            when (it) {
+                                is CredentialStoreState.Success -> {
+                                    token?.let(credentialStoreStateMachine::cancel)
+                                    val accessToken = it.storedCredentials?.cognitoUserPoolTokens?.accessToken
+                                    if (!accessToken.isNullOrEmpty()) {
+                                        runBlocking {
+                                            _fetchDevices(
+                                                accessToken,
+                                                onSuccess,
+                                                onError
+                                            )
+                                        }
+                                    } else {
+                                        onError.accept(AuthException.InvalidStateException())
+                                    }
+                                }
+                                is CredentialStoreState.Error -> {
+                                    token?.let(credentialStoreStateMachine::cancel)
+                                    onError.accept(AuthException.InvalidStateException())
+                                }
+                                else -> {
+                                    // no-op
+                                }
+                            }
+                        },
+                        {
+                            credentialStoreStateMachine.send(
+                                CredentialStoreEvent(CredentialStoreEvent.EventType.LoadCredentialStore())
+                            )
+                        }
+                    )
+                }
+                else -> {
+                    onError.accept(AuthException.SignedOutException())
+                }
+            }
+        }
+    }
+
+    private suspend fun _fetchDevices(
+        token: String,
+        onSuccess: Consumer<MutableList<AuthDevice>>,
+        onError: Consumer<AuthException>
+    ) {
+        try {
+            val response =
+                configureCognitoClients().cognitoIdentityProviderClient?.listDevices(
+                    ListDevicesRequest.invoke {
+                        accessToken = token
+                    }
+                )
+            val _devices = response?.devices
+            val authdeviceList = mutableListOf<AuthDevice>()
+            _devices?.forEach {
+                authdeviceList.add(AuthDevice.fromId(it.deviceKey ?: ""))
```

Comment:
What would `AuthDevice.fromId("")` return?

Reply:
It will just return an instance of AuthDevice("", null) with "" being deviceId and null being devicename. The deviceID can not be null and hence the null guard is added.